AAVAA board addition

aavaa-farnood opened the pull request:
I added the AAVAA V3 board.
Along with the new addition, I had to add more methods for new channel types as well.

Andrey1994 commented:
Could you explain what `rotation_callib_channel` is and how its different from `rotation_channels`?

You can setup this https://brainflow.readthedocs.io/en/stable/BrainFlowDev.html#code-style to fix clang-format CI job

Andrey1994 commented on src/board_controller/aavaa/aavaa_v3.cpp:
```diff
@@ -0,0 +1,563 @@
+        {
+            package[board_descr["default"]["other_channels"][0].get<int> ()] =
+                *reinterpret_cast<uint32_t *> (data_frame + 42) *
+                TIMESTAMP_SCALE; // get_timestamp ();
```
is it a timestamp? If so why is it in other_channels?

aavaa-farnood replied:
This is a board clock info channel for debug purposes, it is different from the `timestamp` channel that stores the timestamps when data is pushed in.

Andrey1994 reviewed (commented):
looks really good!

I am not confident that rotation_calib_channel is really needed, looks too specific as for me, maybe it makes sense to move it to other_channels instead.

Also, I dont fully understand what do you mean by rotation_channels but as for me it looks like data from IMU/gyroscrope sensor, and there is a convention to use `get_gyro_channels` for it, this way it will be convenient with all other devices

aavaa-farnood commented:
> Could you explain what `rotation_calib_channel` is and how its different from `rotation_channels`?
> 
> You can setup this https://brainflow.readthedocs.io/en/stable/BrainFlowDev.html#code-style to fix clang-format CI job

`rotation_calib_channel` can definitely be removed. It is a byte giving info about the calibration of the IMU.

aavaa-farnood commented:
> looks really good!
> 
> I am not confident that rotation_calib_channel is really needed, looks too specific as for me, maybe it makes sense to move it to other_channels instead.
> 
> Also, I dont fully understand what do you mean by rotation_channels but as for me it looks like data from IMU/gyroscrope sensor, and there is a convention to use `get_gyro_channels` for it, this way it will be convenient with all other devices

In the AAVAA board, we have a pitch/roll/yaw info that is different from gyro info and it is processed on the board itself. I realized it's better to create a new set of commands for those channels as it might get confusing.

Andrey1994 commented:
> > Could you explain what `rotation_calib_channel` is and how its different from `rotation_channels`?
> > You can setup this https://brainflow.readthedocs.io/en/stable/BrainFlowDev.html#code-style to fix clang-format CI job
> 
> `rotation_calib_channel` can definitely be removed. It is a byte giving info about the calibration of the IMU.

lets remove it or move this data to other_channels if needed, I am trying to dont create too many methods like this. This way, it is more uniform and also less methods to add when adding support for new programming languages

Andrey1994 commented:
> > looks really good!
> > I am not confident that rotation_calib_channel is really needed, looks too specific as for me, maybe it makes sense to move it to other_channels instead.
> > Also, I dont fully understand what do you mean by rotation_channels but as for me it looks like data from IMU/gyroscrope sensor, and there is a convention to use `get_gyro_channels` for it, this way it will be convenient with all other devices
> 
> In the AAVAA board, we have a pitch/roll/yaw info that is different from gyro info and it is processed on the board itself. I realized it's better to create a new set of commands for those channels as it might get confusing.

ok

Andrey1994 commented:
ok, we can merge it when all ci checks are passed, if you want it before fixing #667 its ok and up to you, or we can fix #667 first

Andrey1994 commented:
Hi!

I still have no mac to take a look at the issue with ble on macos, in the meantime do you have any updates on this? Can merge as soon as all checks are passed

aavaa-farnood commented:
> Hi!
> 
> I still have no mac to take a look at the issue with ble on macos, in the meantime do you have any updates on this? Can merge as soon as all checks are passed and rotation_callib channel is removed

@Andrey1994, I almost gave up after playing around with it for a while.

I added some NSLogs in `centralManagerDidUpdateState` to see the state and I found that it ends up in `CBManagerStatePoweredOn!` five times, and gets stuck there. I added a condition to refuse extra calls after the first one, but did not solve the issue.
I also added the new `WAIT_UNTIL_FALSE_WITH_TIMEOUT` you added but it just makes it not get stuck in that loop.

I don't know what else I can do to solve the issue.

Andrey1994 commented:
if you try to call connect 3 times like in other BLE boards with the latest changes in macos backend to avoid deadlock do all 3 attemps fail?

aavaa-farnood commented:
> if you try to call connect 3 times like in other BLE boards with the latest changes in macos backend to avoid deadlock do all 3 attemps fail?

@Andrey1994, Good news! After doing your suggestion, I can connect and it takes roughly 7 seconds, but I find the behaviour somewhat weird, because it connects to peripheral twice. Here is the NSlog output:

```
2023-09-12 15:07:31.503 Python[61967:1042304] CBManagerStatePoweredOn!
2023-09-12 15:07:31.572 Python[61967:1042304] CBManagerStatePoweredOn!
2023-09-12 15:07:32.667 Python[61967:1042304] CBManagerStatePoweredOn!
2023-09-12 15:07:32.755 Python[61967:1042304] CBManagerStatePoweredOn!
2023-09-12 15:07:32.886 Python[61967:1042326] CBManagerStatePoweredOn!
2023-09-12 15:07:32.903 Python[61967:1040538] Connecting to peripheral...
2023-09-12 15:07:32.903 Python[61967:1040538] Waiting...
2023-09-12 15:07:37.910 Python[61967:1040538] Done waiting.
2023-09-12 15:07:37.910 Python[61967:1040538] Peripheral Connection failed
2023-09-12 15:07:37.911 Python[61967:1040978] Connected to peripheral: <CBPeripheral: 0x12e51a5d0, identifier = 82823F62-BA42-1F50-5027-96A513454000, name = AAVAAB12, mtu = 247, state = connected>
2023-09-12 15:07:38.916 Python[61967:1040538] Connecting to peripheral...
2023-09-12 15:07:38.916 Python[61967:1040538] Waiting...
2023-09-12 15:07:38.917 Python[61967:1042304] Connected to peripheral: <CBPeripheral: 0x12e51a5d0, identifier = 82823F62-BA42-1F50-5027-96A513454000, name = AAVAAB12, mtu = 247, state = connected>
2023-09-12 15:07:38.927 Python[61967:1040538] Done waiting.
2023-09-12 15:07:38.927 Python[61967:1040538] Discovering serives...
```

Andrey1994 commented:
Good, could you do the final cleanup in this PR? Revert changes in simpleble, drop rotation callib channel method, fix clang format job, and check that it builds with --warnings-as-errors like in CI

aavaa-farnood commented:
Sure will do that.
I managed to solve the 7 sec issue, I just needed to add a `sleep(1)` before the first try to connect and it connects in 2 secs.

Andrey1994 commented:
> > if you try to call connect 3 times like in other BLE boards with the latest changes in macos backend to avoid deadlock do all 3 attemps fail?
> 
> @Andrey1994, Good news! After doing your suggestion, I can connect and it takes roughly 7 seconds, but I find the behaviour somewhat weird, because it connects to peripheral twice. Here is the NSlog output:
> 
> ```
> 2023-09-12 15:07:31.503 Python[61967:1042304] CBManagerStatePoweredOn!
> 2023-09-12 15:07:31.572 Python[61967:1042304] CBManagerStatePoweredOn!
> 2023-09-12 15:07:32.667 Python[61967:1042304] CBManagerStatePoweredOn!
> 2023-09-12 15:07:32.755 Python[61967:1042304] CBManagerStatePoweredOn!
> 2023-09-12 15:07:32.886 Python[61967:1042326] CBManagerStatePoweredOn!
> 2023-09-12 15:07:32.903 Python[61967:1040538] Connecting to peripheral...
> 2023-09-12 15:07:32.903 Python[61967:1040538] Waiting...
> 2023-09-12 15:07:37.910 Python[61967:1040538] Done waiting.
> 2023-09-12 15:07:37.910 Python[61967:1040538] Peripheral Connection failed
> 2023-09-12 15:07:37.911 Python[61967:1040978] Connected to peripheral: <CBPeripheral: 0x12e51a5d0, identifier = 82823F62-BA42-1F50-5027-96A513454000, name = AAVAAB12, mtu = 247, state = connected>
> 2023-09-12 15:07:38.916 Python[61967:1040538] Connecting to peripheral...
> 2023-09-12 15:07:38.916 Python[61967:1040538] Waiting...
> 2023-09-12 15:07:38.917 Python[61967:1042304] Connected to peripheral: <CBPeripheral: 0x12e51a5d0, identifier = 82823F62-BA42-1F50-5027-96A513454000, name = AAVAAB12, mtu = 247, state = connected>
> 2023-09-12 15:07:38.927 Python[61967:1040538] Done waiting.
> 2023-09-12 15:07:38.927 Python[61967:1040538] Discovering serives...
> ```

I think its the same for other devices and maybe its the reason why I call it several times, idk much about macos backend in simpleble. Maybe @kdewald can suggest smth.

As long as it works we are good to go, if it calls smth a couple of times but does not hang/crash/etc, its ok for me

aavaa-farnood commented:
> Good, could you do the final cleanup in this PR? Revert changes in simpleble, drop rotation callib channel method, fix clang format job, and check that it builds with --warnings-as-errors like in CI

I did all of those, except for when I run `build.py` with `--warnings-as-errors` it has a lot of `deprecation ` warnings related to `sprintf` and they result in errors. Was not part of my code so didn't touch that.

Andrey1994 commented:
> I did all of those, except for when I run build.py with --warnings-as-errors it has a lot of deprecation  warnings related to sprintf and they result in errors. Was not part of my code so didn't touch that.

Thanks, its fine, I worry about CI job and seems like there its not a warning, maybe you have a newer version of compiler locally.

I see there is a still failing clang-format job, I will fix it by myself, for the future you need to configure this - https://brainflow.readthedocs.io/en/stable/BrainFlowDev.html#code-style with the same clang-format version as in CI job.

I will create a new release tomorrow, so your new board will be available using brainflow installed from package managers.

Andrey1994 commented:
Couple of final questions:

1. Do you want your board to be visible here - https://brainflow.readthedocs.io/en/stable/SupportedBoards.html and do you want me to write a blog post or smth at brainflow.org? In that case I will need some images and links, you can also send a PR to update docs.
2. I expect there will be changes in simpleble in the future and some validation will be needed, I have no this device, so can I ask you to run some tests? BrainFlow slack is a better place for such stuff, so will be good if you ping me there
3. If you are working on it on behalf of a company I will be grateful if your company helps to promote brainflow a little bit)
